pkg/schema/searchComplete.go

```diff
@@ -2,61 +2,104 @@ package schema
-		}
-		if limit != nil {
-			limitStr = strconv.Itoa(*limit)
+func (s *SearchCompleteResult) autoComplete(ctx context.Context) ([]*string, error) {
+	query, args := s.searchCompleteQuery(ctx)
+	res, autoCompleteErr := s.searchCompleteResults(query, args)
+	if autoCompleteErr != nil {
+		klog.Error("Error resolving properties in autoComplete", autoCompleteErr)
+	}
+	return res, autoCompleteErr
+}
+
+func SearchComplete(ctx context.Context, property string, srchInput *model.SearchInput, limit *int) ([]*string, error) {
+	var searchCompleteResult *SearchCompleteResult
+
+	if srchInput != nil {
+		searchCompleteResult = &SearchCompleteResult{
+			input:    srchInput,
+			pool:     db.GetConnection(),
+			property: property,
+			limit:    limit,
 		}
+	}
+	return searchCompleteResult.autoComplete(ctx)
+
+}
 
-		if limitStr != "0" && limitStr != "" {
-			limitClause = " LIMIT " + limitStr
-			query = selectClause + limitClause
+func (s *SearchCompleteResult) searchCompleteQuery(ctx context.Context) (string, []interface{}) {
+	var limit int
+	var whereDs []exp.Expression
+	var selectDs *goqu.SelectDataset
 
+	//FROM CLAUSE
+	schemaTable := goqu.S("search").Table("resources")
+	ds := goqu.From(schemaTable)
+	//WHERE CLAUSE
+	whereDs = res.WhereClauseFilter(s.input)
+	//SELECT CLAUSE
+	if s.property != "" {
+		if s.property == "cluster" {
+			selectDs = ds.SelectDistinct(s.property).Order(goqu.C(s.property).Asc())
+			//Adding notNull clause to filter out NULL values and ORDER by sort results
+			whereDs = append(whereDs, goqu.C(s.property).IsNotNull())
+		} else {
+			selectDs = ds.SelectDistinct(goqu.L(`"data"->>?`, s.property)).Order(goqu.L(`"data"->>?`, s.property).Asc())
+			//Adding notNull clause to filter out NULL values and ORDER by sort results
+			whereDs = append(whereDs, goqu.L(`"data"->>?`, s.property).IsNotNull())
+		}
+		//LIMIT CLAUSE
+		if s.input.Limit != nil && *s.input.Limit != 0 {
+			limit = *s.input.Limit
 		} else {
-			query = selectClause
+			limit = config.DEFAULT_QUERY_LIMIT
+		}
+		//Get the query
+		sql, params, err := selectDs.Where(whereDs...).Limit(uint(limit)).ToSQL()
+		if err != nil {
+			klog.Infof("Error building SearchComplete query: %s", err.Error())
 		}
-		klog.Info("SearchComplete Query: ", query)
-		return query
+		klog.Infof("SearchComplete: %s\nargs: %s", sql, params)
+		return sql, params
 	}
 
-	return ""
+	return "", nil
 }
 
-func searchCompleteResults(query string) ([]*string, error) {
+func (s *SearchCompleteResult) searchCompleteResults(query string, args []interface{}) ([]*string, error) {
 
-	pool := db.GetConnection()
 	//TODO: Handle error
-	rows, _ := pool.Query(context.Background(), query)
+	rows, err := s.pool.Query(context.Background(), query, args...)
+	if err != nil {
+		klog.Error("Error fetching results from db ", err)
+	}
 	defer rows.Close()
 	var srchCompleteOut []*string
-	prop := ""
+	var prop string
```

Comment:
- define the variable prop inside the for loop.
- Then you can delete this line. 
  ```
  tmpProp := prop
  ```

Reply:
Done